### ISSUE TITLE
receiver's latest snapshot is not present on sender 

### ISSUE THREAD

**Issue by lapo-luchini** [zrepl/zrepl#248]
I am using FreeBSD's package `zrepl-0.1.1` configured exactly as [in the tutorial](https://zrepl.github.io/tutorial.html).
I had a blackout on the "sink" side, the host came back up, and then the "push" side shown this:

```
Job: backup                                                                                                                                                                                             
    Type: push                                                                                                                                                                                          
    Replication:                                                                                                                                                                                        
        Attempt #1                                                                                                                                                                                      
        Status: filesystem-error                                                                                                                                                                        
        Problem: one or more of the filesystems encountered errors                                                                                                                                      
        Progress: [==================================================>] 0 B / 0 B @ 0 B/s                                                                                                               
          z/data         PLANNING-ERROR (step 0/0, 0 B/0 B)                                                                                                                                             
            the receiver's latest snapshot is not present on sender:                                                                                                                                    
              last common: @zrepl_20191126_174207_000                                                                                                                                                   
              sender-only:                                                                                                                                                                              
                @zrepl_20191126_175207_000                                                                                                                                                              
                @zrepl_20191126_180207_000                                                                                                                                                              
                @zrepl_20191126_181207_000                                                                                                                                                              
                @zrepl_20191126_182207_000                                                                                                                                                              
                @zrepl_20191126_183207_000                                                                                                                                                              
                @zrepl_20191126_184207_000                                                                                                                                                              
                @zrepl_20191126_185207_000
                (…)
```
I'm not sure how to read that, but it seems that either the "send" side pruned one snapshot too many, or the receiving side lost the last one in the blackout?
Is there any way to recover this situation?
Wht's the best course of action to avoid it in the future?
Using a `last_n` value much higher than `10` (as in the tutorial), maybe? (if it is longer than the blackout, it should be possible to keep the send-but-lost snapshot, assiming that's what happened? I'm not sure)

**Comment by lapo-luchini** [comment]
Uh, no, I think I understand: the "sink" side is a FreeNAS and has automatic (local) snapshots too, so after a few `zrepl_YYYYMMDD*` snapshot there is a single `auto-YYYYMMDD*` snapshot which is probably the "latest" not present on sender.
Can't the sender be instructed to just ignore any snapshot not starting with `zrepl_`?

**Comment by problame** [comment]
This kind of feature was present in an early prototype and got removed before 0.1 IIRC.
I think the reasoning was that although _generally_, users only want to replicate automatically created zrepl-made snapshots,   _sometimes_, for example before a system upgrade, users want to manually take a snapshot that is not part of the zrepl snapshot management cycle (i.e. also not affected by pruning).

The functionality was integrated such that the replication logic was only presented the filtered view of the sender's snapshots.
Now, when a sender changed their filter to include manually taken snapshots, the replication logic would be ... confused ... because suddenly there were snapshots on the sender that weren't there before, and that were older than the last successful replication.

=> **Implementation**

- change replication logic to only consider sender's and receiver's snapshot lists from the point of their most recent common snapshot
- wire filtering to `ListFilesystemVersions` in Sender https://github.com/zrepl/zrepl/blob/ff57a73440912c870cc5e5fb8b62b7568f9fbf5b/endpoint/endpoint.go#L84 
  - [ ] What about the Receiver? Should it be able to present a filtered list of snapshots to the sender? At the very least, its most recent snapshot must always be included, otherwise the replication logic gets a false impression of what's possible to receive
- config wiring
- docs

Would you like to provide a config syntax proposal for snapshot filtering during replication? Please think about other potential use cases and maybe include documenting comments / phrasing that could be included in the official zrepl docs.
For the config syntax, please be aware that the ongoing work on encrypted + resumable send & recv will add a SenderOptions and ReceiverOptions field in the job configs ( see https://github.com/zrepl/zrepl/blob/b1bd9d495cf847a49874b2d536d69cddcea2a6fd/config/config.go#L78-L80 and https://github.com/zrepl/zrepl/blob/ecf990d6f0c5fdefc454e7e4a269755f41bc0716/config/samples/push.yml#L13-L14 )

**Comment by Ranlvor** [comment]
This problem still exsits on v0.3.0-rc1 when using some other tool that manages snapshots, too. Zrepl keeps failing plans because by the time it comes to execute some step of the plan the other tool already deleted the snapshot that was planed to be send, giving an error like
```STEP-ERROR (step 0/26, 0 B/15.8 KiB) validate send arguments: `To` invalid: dataset "rpool@zfs-auto-snap_frequent-2020-07-05-1500" does not exist```
It's not as bad as it was because of Resumable Send and Hold-Protected Send.

Workaround: call `zrepl signal wakeup` till success.

My config syntax proposal would be: no extra config for this feature. When sending snapshots
1. do all resumes possible (to minimise wasted data transfer)
2. Calculate which snapshots would be deleted by the keep_receiver-rules immediately after the job finishes. Leave out these snaphosts when replicating.

That should cover all use-cases for excludes. You just specify via keey_receiver-rules whatever state the receiver should be in at the end of the run and zrepl calculates the minimum amount of sends needed to do that. That would be even beneficial to users without non-zrepl-snapshots. For example if you have hourly snapshots on the sender but only want to save 24 hourlys and 7 dailys on the receiver: Why bother sending all hourlys after an extended downtime when they just would get deleted at the receive side immediately after replication?

**Comment by problame** [comment]
@Ranlvor 

> Calculate which snapshots would be deleted by the keep_receiver-rules immediately after the job finishes. Leave out these snaphosts when replicating.

How would that help if an external tool like `zfs-auto-snapshot` does the pruning?

---

For general robustness, I think my proposal in #349 would be more appropriate.

**Comment by Ranlvor** [comment]
I was thinking about a setup where the external tool manages the snapshots on the sender but zrepl manages the snapshots on the receiver.

**Comment by problame** [comment]
> change replication logic to only consider sender's and receiver's snapshot lists from the point of their most recent common snapshot

This part of the **Implementation** [proposal above](https://github.com/zrepl/zrepl/issues/248#issuecomment-559726992) has been implemented in 3f8fe3a368f4b496758d4cfc1a3edb869d78b6f2 which will be part of 0.3.1.

Is there still interest for sender-side snapshot filtering among the subscribers of this issue?
Would you be willing to test a beta release that implements it?

**Comment by calistoc** [comment]
I think it would be good to have the possibility to filter the snapshots that get replicated
Just as a proposal for the configuration syntax I would add the following optional parameter in the configuration job, either push, pull or source, something like this:

- type: push
  name: to_server2
  replication_snapshots_prefix: zrepl_
…

This would replicate only snapshots with zrepl_ prefix, not specifying this parameter would leave things as they are now replicating all snapshots and preserving the current functionalities
This is just an idea as there might be better ways, like for example using a regex.

Thank you!

**Comment by problame** [comment]
I renamed this issue back to its original title because sender-side snapshot filtering was the solution to some of the symptoms, not the actual problem.

Also, I think we have created more appropriate follow-up issues for each of the use cases / problems encountered

* @Ranlvor's [use case ](https://github.com/zrepl/zrepl/issues/248#issuecomment-653919719) would be handled by https://github.com/zrepl/zrepl/issues/349
* @lapo-luchini 's [problem](https://github.com/zrepl/zrepl/issues/248#issue-529206360) is probably already solved by zrepl 0.3's `replication.protection` option but if snapshot filtering is desired the issue to track is possibly https://github.com/zrepl/zrepl/issues/403

I hope I haven't overlooked anything. If so please comment below.
